simd-support/simd-r5v.h

```diff
@@ -0,0 +1,473 @@
+typedef DS(vuint64m1_t, vuint64m2_t) Vint64;
+typedef DS(vbool64_t, vbool32_t) Vmask;
+#define INT2MASK(imask) DS(__riscv_vmsne_vx_u64m1_b64, __riscv_vmsne_vx_u32m1_b32)(imask, 0, 2*VL)
+
+/* unused */
+/*
+#ifdef FFTW_SINGLE
+#warning "VLIT F32 perhaps should broadcast 64 bits"
+#endif
+static inline V VLIT(const R re, const R im) {
+  Vint idx = TYPEINT(vid_v)(2*VL); // (0, 1, 2, 3, ...)
+  Vint vone = TYPEINT(vmv_v_x)(1, 2*VL);
+  V vre = TYPE(vfmv_v_f)(re, 2*VL);
+  V vim = TYPE(vfmv_v_f)(im, 2*VL);
+  Vint mask = TYPEINT(vand_vv)(idx, vone, 2*VL); // (0, 1, 0, 1, 0, ...)
+  return  TYPE(vmerge_vvm)(vre, vim, INT2MASK(mask), 2*VL);
+}
+*/
+
+#define VLIT1(val) TYPE(vfmv_v_f)(val, 2*VL)
+#define LDK(x) x
+#define DVK(var, val) V var = VLIT1(val)
+#define VZERO VLIT1(DS(0.,0.f))
+/* unused */
+//#define VRONE VLIT(DS(1.,1.f),DS(0.,0.f))
+//#define VCI VLIT(DS(0.,0.f),DS(1.,1.f))
+//#define VCONEMI VLIT(DS(1.,1.f),DS(-1.,-1.f))
+//#define VONE  VLIT1(DS(1.,1.f))
+//#define VMINUSONE VLIT1(DS(-1.,-1.f))
+// generate (all 1, 0, all 1, 0, ...) to split real and imagine parts
+#define VPARTSPLIT TYPEINT(vsub_vx)(TYPEINT(vand_vx)(TYPEINT(vid_v)(2*VL), 1, 2*VL), 1, 2*VL)
+
+static inline V VDUPL(const V x) {
+	Vint idx = TYPEINT(vid_v)(2*VL); // (0, 1, 2, 3, ...)
+	//Vint vnotone = TYPEINT(vmv_v_x)(DS(~1ull,~1), 2*VL);
+	Vint hidx = TYPEINT(vand_vx)(idx, DS(~1ull,~1), 2*VL); // (0, 0, 2, 2, ...)
+	return TYPE(vrgather_vv)(x, hidx, 2*VL);
+}
+static inline V VDUPH(const V x) {
+	Vint idx = TYPEINT(vid_v)(2*VL); // (0, 1, 2, 3, ...)
+	//Vint vone = TYPEINT(vmv_v_x)(1, 2*VL);
+	Vint hidx = TYPEINT(vor_vx)(idx, 1, 2*VL); // (1, 1, 3, 3, ...)
+	return TYPE(vrgather_vv)(x, hidx, 2*VL);
+}
+
+static inline V FLIP_RI(const V x) {
+	Vint idx = TYPEINT(vid_v)(2*VL); // (0, 1, 2, 3, ...)
+	//Vint vone = TYPEINT(vmv_v_x)(1, 2*VL);
+	Vint hidx = TYPEINT(vxor_vx)(idx, 1, 2*VL); // (1, 0, 3, 2, ...)
+	return TYPE(vrgather_vv)(x, hidx, 2*VL);
+}
+
+#define VNEG(a)   TYPE(vfsgnjn_vv)(a,a,2*VL)
+#define VADD(a,b) TYPE(vfadd_vv)(a,b,2*VL)
+#define VSUB(a,b) TYPE(vfsub_vv)(a,b,2*VL)
+#define VMUL(a,b) TYPE(vfmul_vv)(a,b,2*VL)
+#define VFMA(a, b, c)  TYPE(vfmacc_vv)(c,a,b,2*VL) // or vfmadd
+#define VFMS(a, b, c)  TYPE(vfmsac_vv)(c,a,b,2*VL)
+#define VFNMS(a, b, c) TYPE(vfnmsac_vv)(c,a,b,2*VL)  // or vfnmsub
+#define VFMAI(b, c) VADD(c, VBYI(b)) // fixme: improve
+#define VFNMSI(b, c) VSUB(c, VBYI(b)) // fixme: improve
+#define VFMACONJ(b,c)  VADD(VCONJ(b),c) // fixme: improve
+#define VFMSCONJ(b,c)  VFMACONJ(b,VNEG(c)) // fixme: improve
+#define VFNMSCONJ(b,c) VNEG(VFMSCONJ(b,c)) // fixme: improve
+
+#if 1
+static inline V VCONJ(const V x) {
+	Vint idx = TYPEINT(vid_v)(2*VL); // (0, 1, 2, 3, ...)
+	//Vint vone = TYPEINT(vmv_v_x)(1, 2*VL);
+	Vint hidx = TYPEINT(vand_vx)(idx, 1, 2*VL); // (0, 1, 0, 1, 0, 1)
+	//return TYPEMASK(vfsgnjn_vv)(x, x, x, INT2MASK(hidx), 2*VL);
+	return TYPEMERGEDMASK(vfsgnjn_vv)(INT2MASK(hidx), x, x, x, 2*VL);
```

Comment:
The mask register is scarce resource in rvv, and mask undisturbed policy may hinder out-of-order execution. So the unmask version may have benefits.